### PR TITLE
dbcrossbar-ts: Specify schemas with a subset of TypeScript

### PR DESCRIPTION
**Pull request by emk**
We need some way to specify schemas containing complex, nested
structs. This will be particularly useful for BigQuery and for
importing structured JSON data.

You should now be able to write:

    dbcrossbar conv \
      'dbcrossbar-ts:schema.ts#Type' \
      dbcrossbar-schema:type.json

This is implemented using another `peg` parser, which parses
TypeScript using a Parser Expression Grammar (PEG). This is sort of
like a hierarchical regular expression with names rules.

We devote a substantial amount of effort to providing excellent error
messages that highlight errors directly in the TypeScript source.

We support basic types including `interface`, `type` (for aliases),
arrays, primitive types, and `field: ty | null` syntax. We require
semi-colons and commas everywhere they can go, though we may
eventually relax that.

---

**Review Note:** Be sure to expand `ast.ts`; it's where all the interesting bits are.

### CODE REVIEW

File: dbcrossbarlib/src/parse_error.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+        // Build our diagnostic.
+        let diagnostic = Diagnostic::error().with_message(&self.message).with_labels(
+            self.annotations
+                .iter()
+                .map(|a| match a.ty {
+                    AnnotationType::Primary => {
+                        Label::primary(file_id, &a.location).with_message(&a.message)
+                    }
+                    AnnotationType::Secondary => {
```
Comment (seamusabshere) [inline dbcrossbarlib/src/parse_error.rs:59]: this is when you know somebody cares

File: dbcrossbarlib/src/drivers/dbcrossbar_ts/shopify.ts [hunk 2]
```diff
@@ -0,0 +1,329 @@
+// This is a Shopify REST schema that we built by reading the docs.
+
+// A decimal value represented as a string for accuracy.
+type Decimal = string;
+
+// A type that we still need to look up.
+//type TODO = string;
+
+// https://shopify.dev/docs/admin-api/rest/reference/orders/order?api[version]=2020-04
+interface Order {
```
Comment (seamusabshere) [inline dbcrossbarlib/src/drivers/dbcrossbar_ts/shopify.ts:10]: um, wow

Comment (emk) [reply dbcrossbarlib/src/drivers/dbcrossbar_ts/shopify.ts:10]: Yeah, it's pretty complex. There's a reason I built a new schema parser just for this. :-)

This isn't the final schema. We'll need to add lots of `| null`, etc.

**Comment by emk** [comment]
@seamusabshere In theory, this should allow BigQuery -> BigQuery copies that preserve STRUCTs. Or at least it should get us pretty close.

**Pull request merged**